apis/python/src/tiledbsoma/io/registration/signatures.py

```diff
@@ -0,0 +1,308 @@
+import json
+from typing import Dict, Optional
+
+import anndata as ad
+import attrs
+import pandas as pd
+import pyarrow as pa
+from typing_extensions import Self
+
+import tiledbsoma
+import tiledbsoma.logging
+from tiledbsoma._arrow_types import df_to_arrow
+
+
+def _stringify_type(t: pa.DataType) -> str:
+    """
+    Turns an Arrow data type into a stringi more suitable for logging error messages to users in a
+    distributed-computing/distributed-logging environment.
+
+    As noted in the Signature class, we pre-check logic from the ingestor.  As detailed elsewhere,
+    Arrow string and large_string must map to TileDB string, which is large-only. Thus string and
+    large_string form an equivalence class. Similarly for Arrow binary and large_binary.
+    """
+    _EQUIVALENCES = {
+        "large_string": "string",
+        "large_binary": "binary",
+    }
+    str_t = str(t)
+    return _EQUIVALENCES.get(str_t, str_t)
+
+
+def _string_dict_from_arrow_schema(schema: pa.Schema) -> Dict[str, str]:
+    """
+    Converts an Arrow schema to a string/string dict, which is easier on the eyes,
+    easier to convert from/to JSON for distributed logging, and easier to do del-key on.
+    """
+
+    retval = {name: _stringify_type(schema.field(name).type) for name in schema.names}
+
+    # The soma_joinid field is specific to SOMA data but does not exist in AnnData/H5AD.  When we
+    # pre-check an AnnData/H5AD input to see if it's appendable to an existing SOMA experiment, we
+    # must not punish the AnnData/H5AD input for it not having a soma_joinid column in its obs and
+    # var.
+    retval.pop("soma_joinid", None)
+    return retval
+
+
+def _string_dict_from_pandas_dataframe(
+    df: pd.DataFrame,
+    default_index_name: str,
+) -> Dict[str, str]:
+    """
+    Here we provide compatiblity with the ingestor.
+
+    SOMA experiments are indexed by int64 soma_joinid and this is SOMA-only.
+
+    AnnData inputs have a column offered as the index. This can be: named explicitly "obs_id",
+    "var_id", etc.; unnamed: adata.obs.index.name is None; named "index".
+
+    In the latter two cases the ingestor allows a rename to the user's choice
+    such as "obs_id" and "var_id". Here in the appender pre-check logic, we
+    allow the same.
+    """
+
+    df = df.head(1)  # since reset_index can be expensive on full data
+    if df.index.name is None or df.index.name == "index":
+        df.reset_index(inplace=True)
+        df.rename(columns={"index": default_index_name}, inplace=True)
+    else:
+        df.reset_index(inplace=True)
+
+    arrow_table = df_to_arrow(df)
+    arrow_schema = arrow_table.schema.remove_metadata()
+    return _string_dict_from_arrow_schema(arrow_schema)
+
+
+@attrs.define
```

Comment:
I would [use `kw_only` here](https://www.attrs.org/en/stable/api-attr.html#attr.s:~:text=kw_only%20,__init__) to avoid accidentally passing things in as positional arguments. 

```python
@attrs.define(kw_only=True)
class Signature:
  ...
```